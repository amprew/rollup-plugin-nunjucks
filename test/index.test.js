import fs from 'fs';
import fsExtra from 'fs-extra';
import nunjucks from 'nunjucks';
import { rollup } from 'rollup';
import NunjucksPlugin from '../src/index';
import { getAllTemplateFixtures } from './utils/test';

jest.mock('nunjucks', () => ({
  ...jest.requireActual('nunjucks'),
  configure: jest.fn()
}));

process.chdir(`${__dirname}/fixtures`);

async function build(pluginConfig) {
  const bundle = await rollup({
    input: './src/index.js',
    plugins: [NunjucksPlugin(pluginConfig)]
  });

  const meta = await bundle.write({
    format: 'cjs',
    exports: 'auto',
    output: { file: './dist/index.js' }
  });
  await bundle.close();

  const outPath = pluginConfig.output;
  const templateFileContents = await fs.promises.readFile(outPath, 'utf-8');

  return {
    meta,
    fileContent: templateFileContents,
    path: outPath
  };
}

afterEach(async () => {
  if (!fs.existsSync('./dist')) return;

  await fsExtra.remove('./dist');
});

describe('NunjucksPlugin', () => {
  describe('broken config', () => {
    test('when no input specified it throws an error', async () => {
      expect(build()).rejects.toThrowError(/no input file specified$/);
    });

    test('when no output specified it throws an error', async () => {
      expect(build({ input: './src/index.js' })).rejects.toThrowError(
        /no output file specified$/
      );
    });
  });

  test('pass options to nunjucks configuration', async () => {
    await build({
      input: './templates/basic-html.njk',
      output: './dist/index.html',
      opts: {
        test: 'option'
      }
    });

    expect(nunjucks.configure).toHaveBeenCalledWith({
      test: 'option'
    });
  });

  describe('nunjucks templating', () => {
    test('render basic HTML', async () => {
      const { fileContent } = await build({
        input: './templates/basic-html.njk',
        output: './dist/index.html'
      });

      expect(fileContent).toMatchSnapshot();
    });

    test('can loop', async () => {
      const { fileContent } = await build({
        input: './templates/loop.njk',
        output: './dist/index.html'
      });

      expect(fileContent).toMatchSnapshot();
      expect(fileContent).toMatch(/12345/);
    });

    test('if statement', async () => {
      const { fileContent } = await build({
        input: './templates/if-statement.njk',
        output: './dist/index.html'
      });

      expect(fileContent).toMatchSnapshot();
      expect(fileContent).toMatch(/if statement true/);
      expect(fileContent).not.toMatch(/if statement false/);
    });

    test('variables', async () => {
      const { fileContent } = await build({
        input: './templates/variables.njk',
        output: './dist/index.html',
        vars: {
          variable1: 'test variable value',
          listVariable: ['a', 1, 'b', 2, 'c', 3]
        }
      });

      expect(fileContent).toMatchSnapshot();
      expect(fileContent).toMatch(/test variable value/);
      expect(fileContent).toMatch(/3-c-2-b-1-a/);
    });
  });
});
