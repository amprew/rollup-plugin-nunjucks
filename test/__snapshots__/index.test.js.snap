// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NunjucksPlugin nunjucks templating can loop 1`] = `
"12345
"
`;

exports[`NunjucksPlugin nunjucks templating if statement 1`] = `
"
if statement true



"
`;

exports[`NunjucksPlugin nunjucks templating render basic HTML 1`] = `
"<section>
  <div>
    normal html <span class=\\"test\\">elements</span>
  </div>

  <h1>A title here</h1>

  <input type=\\"text\\" name=\\"self-closing-tag\\" />
</section>
"
`;

exports[`NunjucksPlugin nunjucks templating variables 1`] = `
"test variable value
3-c-2-b-1-a
"
`;
